### ISSUE TITLE
[BUG] `CLHHashDelete` fails to delete an element if it's the first one in the linked list.

### ISSUE THREAD

**Issue by boulanlo** [nkallima/sim-universal-construction#51]
**Code version**
Tested on `v3.1.1`, though the bug should still be present in `v3.1.2` ([325fe30](https://github.com/nkallima/sim-universal-construction/commit/325fe30c62ae167dd375b85cb0c86834f400b8e3)).

**Machine architecture/setup**
Tested on `GNU/Linux 5.17.1 x86_64`.

**Compiler**
Using GCC 11.3.0.

**Describe the bug**
Using a `CLHHash` structure, if you issue a `CLHHashDelete()` call for an element that is the first in the underlying linked-list construct, the element isn't deleted at all and nothing happens, leaving the element in place.

This is due to the following snippet in `serialOperations()`:

https://github.com/nkallima/sim-universal-construction/blob/325fe30c62ae167dd375b85cb0c86834f400b8e3/libconcurrent/concurrent/clhhash.c#L75-L80

In the event that we delete the first element, we have:
- `found = true`
- `cur = top`
- `prev = top`

Thus, the line executed is `top->next = cur->next`. Since `top == cur`, this does nothing, and the element stays present.

**To Reproduce**
Compile this code (using the correct linkage and libraries)
```c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <string.h>
#include <stdint.h>

#include <config.h>
#include <primitives.h>
#include <fastrand.h>
#include <threadtools.h>
#include <clhhash.h>
#include <barrier.h>
#include <bench_args.h>
#include <math.h>

#define N_BUCKETS        64
#define LOAD_FACTOR      1
#define INITIAL_CAPACITY (LOAD_FACTOR * N_BUCKETS)

#define RANDOM_RANGE (INITIAL_CAPACITY * log(INITIAL_CAPACITY))

CLHHash object_struct CACHE_ALIGN;
CLHHashThreadState *th_state;

int main(int argc, char *argv[]) {
    CLHHashStructInit(&object_struct, N_BUCKETS, 16);

    th_state = synchGetAlignedMemory(CACHE_LINE_SIZE, sizeof(CLHHashThreadState));
    CLHHashThreadStateInit(&object_struct, th_state, N_BUCKETS, 0);

    CLHHashInsert(&object_struct, th_state, 1, 1, 0);
    CLHHashDelete(&object_struct, th_state, 1, 0);

    int search = CLHHashSearch(&object_struct, th_state, 1, 0);

    fprintf(stderr, "search: %d\n", search);

    return 0;
}
```
In this, we initialize a `CLHHash` structure, insert an element to it, delete it immediately, and search for it afterwards.

**Expected behavior**
Given that we search for an element that has just been deleted, I expect the program to print `search: 0`.

However, the current behavior is to print `search: 1`.

**Comment by nkallima** [comment]
@boulanlo Thanks for reporting this.
I will try to reproduce the bug soon.

**Comment by nkallima** [comment]
@boulanlo 
I successfully reproduced the bug. I have pushed a bug fix in the v3.1.3-dev branch.
Please check if this fix works for you.

Thanks again for your report.

**Comment by boulanlo** [comment]
The fix you pushed in the v3.1.3-dev branch (96052b1a75cda4fed4a1aa01f19e38ff50a1ca5a) resolved the problem. Thanks for your quick answer and fix!
